File: 프로그래머스/lv3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.*;

class Solution {
    public int solution(int[] stones, int k) {
        int answer = 0;
        int min = 1;
        int max = 200000000;
        
        while (min <= max) {
            int mid = (min + max) / 2;
            
            if (canPass(stones, k, mid)) {
                min = mid + 1;
                answer = Math.max(answer, mid);
            } else {
                max = mid - 1;
            }
        }
        return answer;
    }
    
    private boolean canPass(int[] stones, int k, int cnt) {
        int pass = 0;
        
        for (int stone : stones) {
            if (stone - cnt < 0) pass++;
            else pass = 0;
            
            if (pass >= k) return false;
        }
        return true;
    }
}

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> stones, int k) {
    int answer = 0;
    
    int l = 0, r = 200000000;
    
    while (l <= r) {
        vector<int> temp = stones;
        int jump = 0;
        int max_jump = 0;
        int mid = (l + r) / 2;
        for (int i = 0; i < temp.size(); i++) {
            if (temp[i] - mid < 0) {
                jump++;
            } else {
                jump = 0;
            }
            max_jump = max(max_jump, jump);
        }
        
        if (max_jump >= k) {
            r = mid - 1;
        } else {
            answer = max(answer, mid);
            l = mid + 1;
        }
        
    }
    return answer;
}